src/Lock.sol

```diff
@@ -0,0 +1,37 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+pragma solidity ^0.8.24;
+
+import {Locker} from "./libraries/Locker.sol";
+
+/// @title Lock
+/// @notice A contract that provides a reentrancy lock for external calls
+/// @author Modified from https://github.com/Uniswap/universal-router/blob/main/contracts/base/Lock.sol
+contract Lock {
+    /// @notice Thrown when attempting to reenter a locked function from an external caller
+    error ContractLocked();
+
+    /// @notice Modifier enforcing a reentrancy lock that allows self-reentrancy
+    /// @dev If the contract is not locked, use msg.sender as the locker
+    modifier isNotLocked() {
+        // Apply a reentrancy lock for all external callers
+        if (msg.sender != address(this)) {
+            if (Locker.isLocked()) revert ContractLocked();
+            Locker.set(msg.sender);
+            _;
+            Locker.set(address(0));
+        } else if (!Locker.isLocked()) {
+            /// @dev For 7702 accounts the msg.sender is address(this) but we still need to set the lock if unset
+            Locker.set(msg.sender);
+            _;
+            Locker.set(address(0));
```

Comment:
This is interesting, essentially we still need to lock the contract if msg.sender == address(this) for only the top level call, since this is techinically valid under 7702

Reply:
makes sense, I do wonder if we also NEED to prevent external reentrancy? 

any function that authorizes through msgSender() now would fail and any other function authorizes through signatures so IF an external contract wanted to re call into the contract and had a valid signature, what is the worst case here?

Reply:
I think the main outcome should be for any function using _authorizeCaller to require the top level call be initiated by address(this), which this achieves. Do we want to allow _authorize_ via signature? Like, doing a userOp to add a passkey? Like potentially...